winch/codegen/src/isa/x64/asm.rs

```diff
@@ -472,6 +513,47 @@ impl Assembler {
         });
     }
 
+    /// Compare two operands and write bit to dst
+    pub fn cmp(&mut self, src: Operand, dst: Operand, kind: CmpKind, size: OperandSize) {
+        let src = GprMemImm::new(match src {
+            Operand::Reg(r) => r.into(),
+            Operand::Imm(imm) => match i32::try_from(imm) {
+                Ok(i) => RegMemImm::imm(i as u32),
+                Err(_) => {
+                    let scratch = regs::scratch();
+                    self.mov_ir(imm as u64, scratch, size);
+                    scratch.into()
+                }
+            },
+            Operand::Mem(_) => panic!("mem operand for src not supported"),
+        })
+        .expect("valid register or immediate");
+        let dst = match dst {
+            Operand::Reg(r) => r,
+            _ => panic!("dst must be a register"),
+        };
+
+        // sets flags like `ZF` (zero flag) in status register
+        self.emit(Inst::CmpRmiR {
+            size: size.into(),
+            opcode: CmpOpcode::Cmp,
+            src,
+            dst: dst.into(),
+        });
+        // clear the dst register or bits 1 to 31 may be incorrectly set
+        // don't use xor since it updates the status register
+        self.emit(Inst::Imm {
+            dst_size: args::OperandSize::Size32, // always going to be an i32 result
+            simm64: 0,
+            dst: dst.into(),
+        });
+        // copy appropriate status register bit into dst's 0 bit
+        self.emit(Inst::Setcc {
+            cc: kind.into(),
+            dst: dst.into(),
+        });
```

Comment:
Should combining these instructions be handled in the micro assembler or the macro assembler? Zeroing out the `dst` register is an interesting case of this because doing it with no context would use `xor` which would reset the status register flags.